File: kotlin/src/main/kotlin/edax/Perft.kt
package edax

/** Performance test - count leaf nodes by expanding all legal moves to a given depth. */
fun perft(board: Board, depth: Int): ULong {
    if (depth == 0) return 1uL

    val moves = board.getMoves()

    if (moves == 0uL) {
        val passed = board.copy()
        passed.pass()
        if (passed.getMoves() == 0uL) {
            // Game over
            return 1uL
        }
        return perft(passed, depth - 1)
    }

    var count: ULong = 0uL
    var remaining = moves

    while (remaining != 0uL) {
        val sq = remaining.countTrailingZeroBits()
        remaining = remaining and (remaining - 1uL)

        val child = board.copy()
        child.doMove(sq)
        count += perft(child, depth - 1)
    }

    return count
}

/** Expected perft results for the initial position. */
val PERFT_EXPECTED = ulongArrayOf(
    1uL, 4uL, 12uL, 56uL, 244uL, 1396uL, 8200uL, 55092uL, 390216uL
)

File: swift/Sources/EdaxCore/Perft.swift
/// Performance test - count leaf nodes by expanding all legal moves to a given depth.
public func perft(_ board: Board, depth: Int) -> UInt64 {
    if depth == 0 { return 1 }

    let moves = board.getMoves()

    if moves == 0 {
        var passed = board
        passed.pass()
        if passed.getMoves() == 0 {
            // Game over
            return 1
        }
        return perft(passed, depth: depth - 1)
    }

    var count: UInt64 = 0
    var remaining = moves

    while remaining != 0 {
        let sq = remaining.trailingZeroBitCount
        remaining &= remaining - 1

        var child = board
        child.doMove(sq)
        count += perft(child, depth: depth - 1)
    }

    return count
}

/// Expected perft results for the initial position.
public let perftExpected: [UInt64] = [
    1, 4, 12, 56, 244, 1396, 8200, 55092, 390216
]
